tilings/tiling.py

```diff
@@ -145,18 +169,53 @@ def from_perms(
             t = t.add_list_requirement(req_list)
         return t
 
-    @cssmethodtimer("Tiling._fill_empty")
-    def _fill_empty(self) -> None:
-        """Add size one obstructions to cells that are empty."""
-        add = []
-        (i, j) = self.dimensions
-        active_cells = self.active_cells
-        empty_cells = self.empty_cells
-        for x in range(i):
-            for y in range(j):
-                if (x, y) not in active_cells and (x, y) not in empty_cells:
-                    add.append(GriddedPerm.single_cell(Perm((0,)), (x, y)))
-        self._obstructions = tuple(sorted(tuple(add) + self._obstructions))
+    @cssmethodtimer("Tiling._prepare_properties")
+    def _prepare_properties(self) -> None:
+        """
+        Compute _active_cells, _empty_cells, _dimensions, and store them
+        """
+        active_cells = union_reduce(
+            set(ob.pos) for ob in self.obstructions if len(ob) > 1
+        )
+        active_cells.update(
+            *(union_reduce(set(comp.pos) for comp in req) for req in self.requirements)
+        )
+
+        max_row = 0
+        max_col = 0
+        for cell in active_cells:
+            max_col = max(max_col, cell[0])
+            max_row = max(max_row, cell[1])
+        dimensions = (max_col + 1, max_row + 1)
+
+        empty_cells = tuple(
+            cell
+            for cell in product(range(dimensions[0]), range(dimensions[1]))
+            if cell not in active_cells
+        )
+
+        # If the first obstruction is the empty perm, we shouldn't do any of this.
+        if len(self._obstructions) > 0 and len(self._obstructions[0]) > 0:
+            # We can assume that self._obstructions is sorted at this point, so to
+            #   extract the point obstructions, we just pass though them until we've
+            #   found the last one, then we slice the list there.
+            index = 0
+            for ob in self._obstructions:
+                if len(ob) > 1:
+                    break
+                index += 1
```

Comment:
We should use the next function for this:
```python
index = next(i for i, ob in enumerate(self._obstructions) if len(ob) > 1)
```
I'll make this change

Reply:
Is it fair to assume that there is at least one of size > 1 ?

Reply:
the answer was no, needed to add default value being the length of the obstructions

Reply:
I was worried a solution with enumerate would be slower, and this is one of those things we do millions of times. And yes, you need a default for the case where all obstructions are point obstructions (e.g., the empty tiling).